### ISSUE TITLE
The program freezes after a few minutes of running.

### ISSUE THREAD

**Issue by phamtandatx** [KKest/ESP-rplidarS2#1]
I connect to the sensor and run it, everything seems fine, but after running for a few minutes, the code freezes.

**Comment by mmadyda** [comment]
I have the same problem

**Comment by mmadyda** [comment]
Code freeze after 10 seconds

**Comment by keiaiendiel** [comment]
Hi, has anyone managed to get this working? I've been trying this for a few days and still can't seem to get any data from RPLIDAR S2 to ESP32.. I noticed it freezes when RX pin is connected to RX of ESP, but i did not find any solution using hardware Serials..

**Comment by KKest** [comment]
Hello, 
normaly you have to connect the TX of RPLIDAR to the RX of the ESP32 and the RX of RPLIDAR to the TX of the ESP32. 
I had problems, back when i wrote this code, with the connection to the Sensor due to loose connections. So verify the connection is good (use suitable connectors and cables). 
The code is waiting for a response from the RPLIDAR. If the ESP doesn't get it, the code will freeze.

As this is an old project, the implemented functionality is somewhat limited. Since I don't own a sensor, I can't test it right now. Sorry about that!

**Comment by keiaiendiel** [comment]
Thank you for your quick reply. It is not freezing now, but I only get 0mm serial prints. Should I also connect the MOTOCTRL pin? It was not defined in the example file so I am not sure about that.

**Comment by keiaiendiel** [comment]
actually, i have to say that it finally started turning, so the motor pin is not needed anyway. but the code still prints 0 values, even if it is connected or not. (I get Lidar timeout message) I'm looking at it in the Arduino IDE and using 2 Serials because it wasn't working turning/working before:

rpLidar lidar(&Serial2, 1000000);

void setup() {
  Serial.begin(115200);
  Serial2.begin(1000000, SERIAL_8N1, 2, 4);

**Comment by KKest** [comment]
I don't know which Board you're using but on a Esp32 dev kit is Serial2 at pins GPIO16 and GPIO17. The Hardware Serial gets initialized inside the Library, with the line:
```
rpLidar lidar(&Serial2,1000000);
```

So if you connect the Lidar to GPIO16 and  GPIO17 you have to delete the line:
```
Serial2.begin(1000000,SERIAL_8N1,2,4);
```

Pin 2 and 4 are used for internal Flash from the esp when I remember right. That could lead to errors.

**Comment by keiaiendiel** [comment]
Thanks, I managed to get it working. It was probably really a wiring error. I'm glad I'm finally getting lidar data, but I'm now facing the same problem as my colleagues above. For a short time the code gets stuck and the lidar doesn't start sending data again. I understand that this is not a serial port problem, the jam also occurs when sending data via OSC for example. Anyway, I can see a difference in the time of the jam depending on how much data I send and how fast.

Isn't there any way to implement a function in the code that notices the connection failure, and tries to reconnect the lidar? Because everything continues after the hit, only the connection to the lidar doesn't work anymore. I don't really understand how to use .deviceHealtStatus etc. I would be very grateful for help.

**Comment by keiaiendiel** [comment]
Thanks, I managed to get it working. It was probably really a wiring error. I'm glad I'm finally getting lidar data, but I'm now facing the same problem as my colleagues above. For a short time the code gets stuck and the lidar doesn't start sending data again. I understand that this is not a serial port problem, the jam also occurs when sending data via OSC for example. Anyway, I can see a difference in the time of the jam depending on how much data I send and how fast.

Isn't there any way to implement a function in the code that notices the connection failure, and tries to reconnect the lidar? Because everything continues after the hit, only the connection to the lidar doesn't work anymore. I don't really understand how to use .deviceHealtStatus etc. I would be very grateful for help. My code doesn't continue after this line : 
`stDeviceStatus_t status = lidar.getDeviceHealth();`

**Comment by KKest** [comment]
Nice to hear that you receive some data.
The code should receive the Cabin Packets like they´re described in the Protocol of the Lidar. So the start of a new Cabin Packet is detected through the Sync1 and Sync2 bytes with their crc .
The code awaits 2 bytes and checks if the received Bytes are Sync1 with it´s CRC and then checks for the next Sync with CRC . I found a problem when the readout of the Sensor is paused while some other code gets executed. It is possible that the received 2 Bytes are not received as wanted like:
|Received Byte|Received Byte 4 HIGH Bits |Received Byte 4 LOW Bits|
|----|---------|-----------|
|1|Sync1|CRC|
|2|Sync2|CRC|

but are received like:
|Received Byte|Received Byte 4 HIGH Bits |Received Byte 4 LOW Bits|
|----|---------|-----------|
|1|Sync2|CRC|
|2|start_Angle_q6[7:4]|start_Angle_q6[3:0]|

That the Sync1 byte is not found and the code not contuinue. 
There is a solution for this problem and I want to fix it soon.

**Device Health**

```
stDeviceStatus_t status = lidar.getDeviceHealth();
```

The Device Health data will be saved in "status". So you can print the Data with:

```
lidar.DebugPrintDeviceErrorStatus(status);
```
|status|Information|
|----|---------|
|0|Good|
|1|Warning|
|2|Error|

I did not found any information about the errorcode-bytes in the protocol datasheet.

**Comment by keiaiendiel** [comment]
Hey, thanks. I found some sufficient way when the lidar restarts to prevent freezing. Unfortunately, I don't really see much into the sync issue, but I wish you good luck in solving it. I have one more question 🙈 I noticed that the scan range is defined in the code.

`lidar.setAngleOfInterest(5,175);`

Is it possible to increase it to the whole sensor circuit (360° )? Whenever it exceeds 180 degrees (also 100-280; doesn't have to go from zero), it stops sending data.

**Comment by ElbyB** [comment]
Hi,

I'm also trying to use a RPLIDAR S2 with esp32 (WESP32 rev 7) and I'm having the same freezing issue. The freezing happens randomly between 2 and 70 reading loop. Here is my code :


```cpp

#include "rpLidar.h"
#include "rpLidarTypes.h"
#include <ETH.h>
#include <OSCMessage.h>
#include <OSCBundle.h>
#include <OSCData.h>
#include <WiFiUdp.h>
WiFiUDP Udp;
rpLidar lidar(&Serial1,1000000);

//#define RPLIDAR_MOTOR 18 // The PWM pin for control the speed of RPLIDAR's motor.


const IPAddress ipEth(192, 168, 1, 203);
const IPAddress gateway(192, 168, 1, 1);
const IPAddress subnet(255, 255, 255, 0);

static bool eth_connected = false;

// for OSC
unsigned int receivePort = 8888;
IPAddress outIp(192, 168, 1, 201);
unsigned int outPort = 7003;

// Counters for milliseconds during interval
long previousMillis = 0;
long currentMillis = 0;
int interval = 2000;

int distance = 0;
float angle = 0;


void setup() {
  delay(2000);//wait for the controlleur to be ready

// Serial 0 (console)
  //Serial.begin(115200);
  //Serial1.begin(115200);

  lidar.resetDevice(); //reset the device to be sure that the status is good
  lidar.setAngleOfInterest(5,175); //Set the field of view that is saved to Data
  lidar.start(express); //start the express scan of the lidar

// ETH
    WiFi.onEvent(WiFiEvent);
    ETH.begin(0, -1, 16, 17, ETH_PHY_RTL8201);
    //ETH.begin();
    ETH.config(ipEth, gateway, subnet);

}

void loop() {
  
  //analogWrite(RPLIDAR_MOTOR, 0);
  
  lidar.readMeasurePoints();// reads a full scan and save it to Data
  for(int i=0;i<sizeof(lidar.Data)/sizeof(point_t);i++)
  {
    /*Serial.print(lidar.Data[i].angle);
    Serial.print("deg\t|\t");
    Serial.print(lidar.Data[i].distance);
    Serial.println("mm");*/

    distance = lidar.Data[i].distance;
    angle = lidar.Data[i].angle;

      OSCMessage msgDist("/lidar");
      msgDist.add(distance);
      Udp.beginPacket(outIp, outPort);
      msgDist.send(Udp);
      Udp.endPacket();
      msgDist.empty();

      OSCMessage msgAngle("/angle");
      msgAngle.add(angle);
      Udp.beginPacket(outIp, outPort);
      msgAngle.send(Udp);
      Udp.endPacket();
      msgAngle.empty();

      
  }
  //delay (500);

  //stDeviceStatus_t status = lidar.getDeviceHealth();
  //lidar.DebugPrintDeviceErrorStatus(status);

  
}

```

Has anyone got it working?

**Comment by KKest** [comment]
Hello, 
unfortunately I don't have any hardware to test with.
I modified the code to prevent it from running in an endless loop.

Can you test it and give me feedback?
Please test the Branch [Test Branch](https://github.com/KKest/ESP-rplidarS2/tree/fix_Issue%231).

**Comment by ElbyB** [comment]
Hi,

Thanks for this! There was some mistakes in rpLidar.cpp where it was written "if(millis()internTimeCount+500)" instead of "if(millis()>internTimeCount+500)", but I changed them and now it is streaming for one hour without a problem!

But now if I change "lidar.setAngleOfInterest(5,175);" to "lidar.setAngleOfInterest(0,360);" it doesn't work anymore.  Is there a way to get all 360 degrees of data?

LR

**Comment by KKest** [comment]
Hi, 

good to hear you got it working.
I fixed the symbol mistake "<" you mentioned.

Does the ESP32 restart every scan cycle of the lidar?
I think the size of the array ExpressDataBuffer has to be increased to store a full 360° dataset.
I put everything together in the commit #[43b6a11](https://github.com/KKest/ESP-rplidarS2/commit/43b6a11da9ffe07675d43289daf2d50431845e4a)
Please test again and give me feedback :)

**Comment by ElbyB** [comment]
Hi,

Yes its doing the tricks thanks. How can I know if  the ESP32 restart every scan cycle of the lidar? And how it is important?

**Comment by KKest** [comment]
fixed with commit [a05aca7](https://github.com/KKest/ESP-rplidarS2/commit/a05aca71703ee9843ac46042d925c08db06b86b2)
